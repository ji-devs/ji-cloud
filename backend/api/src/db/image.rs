use super::{recycle_metadata, recycle_tags};
use chrono::{DateTime, Utc};
use futures::stream::BoxStream;
use shared::domain::{
    category::CategoryId,
    image::{ImageId, ImageKind, ImageMetadata},
    meta::{AffiliationId, AgeRangeId, ImageStyleId, ImageTagIndex},
};
use sqlx::{PgConnection, PgPool};
use uuid::Uuid;

pub mod recent;
pub mod tag;
pub mod user;

pub async fn create(
    conn: &mut PgConnection,
    name: &str,
    description: &str,
    is_premium: bool,
    publish_at: Option<DateTime<Utc>>,
    kind: ImageKind,
) -> sqlx::Result<ImageId> {
    let id: ImageId = sqlx::query!(
        r#"
insert into image_metadata (name, description, is_premium, publish_at, kind) values ($1, $2, $3, $4, $5)
returning id as "id: ImageId"
        "#,
        name,
        description,
        is_premium,
        publish_at,
        kind as i16,
    )
    .fetch_one(&mut *conn)
    .await?
    .id;

    sqlx::query!("insert into image_upload (image_id) values($1)", id.0)
        .execute(&mut *conn)
        .await?;

    Ok(id)
}

pub async fn update_metadata(
    conn: &mut PgConnection,
    image: ImageId,
    affiliations: Option<&[AffiliationId]>,
    age_ranges: Option<&[AgeRangeId]>,
    image_styles: Option<&[ImageStyleId]>,
    categories: Option<&[CategoryId]>,
    image_tags: Option<&[ImageTagIndex]>,
) -> sqlx::Result<()> {
    const TABLE: &str = "image";

    if let Some(affiliations) = affiliations {
        recycle_metadata(&mut *conn, TABLE, image.0, affiliations).await?;
    }

    if let Some(age_ranges) = age_ranges {
        recycle_metadata(&mut *conn, TABLE, image.0, age_ranges).await?;
    }

    if let Some(styles) = image_styles {
        recycle_metadata(&mut *conn, TABLE, image.0, styles).await?;
    }

    if let Some(tags) = image_tags {
        recycle_tags(&mut *conn, TABLE, image.0, tags).await?;
    }

    if let Some(categories) = categories {
        recycle_metadata(&mut *conn, TABLE, image.0, categories).await?;
    }

    Ok(())
}

pub async fn update(
    conn: &mut PgConnection,
    id: ImageId,
    name: Option<&str>,
    description: Option<&str>,
    is_premium: Option<bool>,
    publish_at: Option<Option<DateTime<Utc>>>,
) -> anyhow::Result<bool> {
    if !sqlx::query!(
        r#"select exists(select 1 from image_metadata where id = $1) as "exists!""#,
        id.0
    )
    .fetch_one(&mut *conn)
    .await?
    .exists
    {
        return Ok(false);
    }

    if let Some(publish_at) = publish_at {
        sqlx::query!(
            r#"
update image_metadata
set publish_at = $2, updated_at = now()
where id = $1 and $2 is distinct from publish_at"#,
            id.0,
            publish_at
        )
        .execute(&mut *conn)
        .await?;
    }

    if let Some(description) = description {
        sqlx::query!(
            r#"
update image_metadata
set description = $2,
    translated_description = '{}',
    updated_at = now()
where id = $1 and $2 is distinct from description"#,
            id.0,
            description,
        )
        .execute(&mut *conn)
        .await?;
    }

    sqlx::query!(
        //language=SQL
        r#"
update image_metadata
set name        = coalesce($2, name),
    is_premium  = coalesce($3, is_premium),
    updated_at  = now()
where id = $1
  and (($2::text is not null and $2 is distinct from name) or
       ($3::boolean is not null and $3 is distinct from is_premium))"#,
        id.0,
        name,
        is_premium,
    )
    .execute(conn)
    .await?;

    Ok(true)
}

pub async fn get_one(db: &PgPool, id: ImageId) -> sqlx::Result<Option<ImageMetadata>> {
    sqlx::query_as(
r#"
select id,
       name,
       kind,
       description,
       translated_description,
       is_premium,
       publish_at,
       created_at,
       updated_at,
       translated_description,
       array((select row (category_id) from image_category where image_id = id))       as categories,
       array((select row (style_id) from image_style where image_id = id))             as styles,
       array((select row (age_range_id) from image_age_range where image_id = id))     as age_ranges,
       array((select row (affiliation_id) from image_affiliation where image_id = id)) as affiliations,
       array((select row (tag_index) from image_tag_join where image_id = id))         as tags
from image_metadata
         join image_upload on id = image_id
where id = $1
  and processing_result is true
"#)
    .bind(id)
    .fetch_optional(db)
    .await
}

pub fn list(
    db: &PgPool,
    is_published: Option<bool>,
    kind: Option<ImageKind>,
    page: i32,
    page_limit: u32,
) -> BoxStream<'_, sqlx::Result<ImageMetadata>> {
    sqlx::query_as(
r#"
select id,
       kind,
       name,
       description,
       translated_description,
       is_premium,
       publish_at,
       created_at,
       updated_at,
       array((select row (category_id) from image_category where image_id = id))       as categories,
       array((select row (style_id) from image_style where image_id = id))             as styles,
       array((select row (age_range_id) from image_age_range where image_id = id))     as age_ranges,
       array((select row (affiliation_id) from image_affiliation where image_id = id)) as affiliations,
       array((select row (tag_index) from image_tag_join where image_id = id))         as tags
from (image_metadata
         inner join image_upload on image_id = id)
where processing_result is not distinct from true 
    and (publish_at < now() is not distinct from $1 or $1 is null)
    and (kind is not distinct from $3 or $3 is null) 
order by id
limit $4 offset ($4 * $2)
"#)
    .bind(is_published)
    .bind(page)
    .bind(kind.map(|it| it as i16))
    .bind(page_limit as i16)
    .fetch(db)
}

pub async fn filtered_count(
    db: &PgPool,
    is_published: Option<bool>,
    kind: Option<ImageKind>,
) -> sqlx::Result<u64> {
    sqlx::query!(
        r#"
select count(*) as "count!: i64" 
from image_metadata
        inner join image_upload on image_id = id 
where processing_result is not distinct from true 
    and (publish_at < now() is not distinct from $1 or $1 is null)
    and (kind is not distinct from $2 or $2 is null)"#,
        is_published,
        kind.map(|it| it as i16),
    )
    .fetch_one(db)
    .await
    .map(|it| it.count as u64)
}

pub fn get<'a>(db: &'a PgPool, ids: &'a [Uuid]) -> BoxStream<'a, sqlx::Result<ImageMetadata>> {
    sqlx::query_as(
r#"
select id,
       kind,
       name,
       description,
       translated_description,
       is_premium,
       publish_at,
       created_at,
       updated_at,
       array((select row (category_id) from image_category where image_id = id))       as categories,
       array((select row (style_id) from image_style where image_id = id))             as styles,
       array((select row (age_range_id) from image_age_range where image_id = id))     as age_ranges,
       array((select row (affiliation_id) from image_affiliation where image_id = id)) as affiliations,
       array((select row (tag_index) from image_tag_join where image_id = id))         as tags
from image_metadata
         inner join image_upload on image_id = id
         inner join unnest($1::uuid[])
    with ordinality t(id, ord) USING (id)
where processing_result is not distinct from true
order by ord
"#)
    .bind(ids)
    .fetch(db)
}

pub async fn delete(db: &PgPool, image: ImageId) -> sqlx::Result<()> {
    let mut conn = db.begin().await?;

    // first, clear any metadata it might have.
    update_metadata(
        &mut conn,
        image,
        Some(&[]),
        Some(&[]),
        Some(&[]),
        Some(&[]),
        Some(&[]),
    )
    .await?;

    sqlx::query!("delete from image_upload where image_id = $1", image.0)
        .execute(&mut conn)
        .await?;

    // then drop.
    sqlx::query!("delete from image_metadata where id = $1", image.0)
        .execute(&mut conn)
        .await?;

    conn.commit().await
}
