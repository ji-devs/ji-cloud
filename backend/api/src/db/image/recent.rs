use crate::error;
use chrono::{DateTime, Utc};
use shared::{
    domain::{
        image::{recent::UserRecentImageResponse, ImageId},
        user::UserId,
    },
    media::MediaLibrary,
};
use sqlx::PgPool;

pub async fn upsert(
    db: &PgPool,
    user_id: UserId,
    image_id: ImageId,
    library: MediaLibrary,
) -> anyhow::Result<(ImageId, MediaLibrary, DateTime<Utc>, bool), error::UserRecentImage> {
    let mut txn = db.begin().await?;

    let exists = sqlx::query!(
        r#"
select exists(select 1 from user_recent_image where user_id = $1 and image_id = $2) as "exists!"
            "#,
        user_id.0,
        image_id.0,
    )
    .fetch_one(&mut txn)
    .await?
    .exists;

    let res = sqlx::query!(
            // language=SQL
        r#"
insert into user_recent_image (user_id, image_id, media_library)
values ($1, $2, $3)
ON CONFLICT (user_id, image_id) DO UPDATE
  SET user_id = $1,
    image_id = $2,
    media_library = $3,
    last_used = now()
returning image_id as "id: ImageId", media_library as "library: MediaLibrary", last_used as "last_used: DateTime<Utc>";
        "#,
        user_id.0,
        image_id.0,
        library as i16,
    )
    .fetch_one(&mut txn)
    .await?;

    txn.commit().await?;

    return Ok((res.id, res.library, res.last_used, exists));
}

pub async fn delete(db: &PgPool, user_id: UserId, image_id: ImageId) -> sqlx::Result<()> {
    sqlx::query!(
        r#"
delete from user_recent_image
where user_id = $1 and image_id = $2
            "#,
        user_id.0,
        image_id.0
    )
    .execute(db)
    .await?;

    Ok(())
}

pub async fn list(
    db: &PgPool,
    user_id: UserId,
    limit: Option<i64>,
) -> sqlx::Result<Vec<UserRecentImageResponse>> {
    // if let Some(limit) = limit { assert!(limit >= 0); }

    sqlx::query_as!(
            UserRecentImageResponse,
            r#"
select image_id as "id: ImageId", media_library as "library: MediaLibrary", last_used as "last_used: DateTime<Utc>"
from user_recent_image
where user_id = $1
order by last_used desc
limit $2
            "#,
            user_id.0,
            limit
        )
        .fetch_all(db)
        .await
}
