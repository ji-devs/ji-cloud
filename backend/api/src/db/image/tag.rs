use crate::error;
use shared::domain::{image::tag::ImageTagResponse, meta::ImageTagIndex};
use sqlx::{postgres::PgDatabaseError, PgPool};

pub fn handle_tag_err(err: sqlx::Error) -> error::Tag {
    match err {
        sqlx::Error::Database(err)
            if err.downcast_ref::<PgDatabaseError>().constraint()
                == Some("image_tag_index_pkey") =>
        {
            error::Tag::TakenIndex
        }
        e => e.into(),
    }
}

pub async fn list(db: &PgPool) -> sqlx::Result<Vec<ImageTagResponse>> {
    sqlx::query_as!(
        ImageTagResponse,
        r#"
select index as "index: ImageTagIndex", display_name from "image_tag"
order by index
            "#
    )
    .fetch_all(db)
    .await
}

pub async fn create(
    db: &PgPool,
    index: ImageTagIndex,
    display_name: &str,
) -> Result<(ImageTagIndex, String), error::Tag> {
    let mut txn = db.begin().await?;

    let res = sqlx::query!(
        // language=SQL
        r#"
insert into image_tag (index, display_name)
values ($1, $2)
returning index as "index: ImageTagIndex", display_name
            "#,
        index.0,
        display_name,
    )
    .fetch_one(&mut txn)
    .await
    .map_err(handle_tag_err)?;

    txn.commit().await?;

    Ok((res.index, res.display_name))
}

pub async fn update(
    db: &PgPool,
    curr_index: i16,
    display_name: Option<&str>,
    new_index: Option<i16>,
) -> Result<(), error::Tag> {
    let mut txn = db.begin().await?;

    let res = sqlx::query!(
        r#"select index as "index: i16" from image_tag where index = $1 for update"#,
        curr_index
    )
    .fetch_optional(&mut txn)
    .await?;

    if res.is_none() {
        txn.commit().await?;
        return Err(error::Tag::ResourceNotFound);
    }

    if let Some(display_name) = display_name {
        sqlx::query!(
            r#"update image_tag set display_name = $2 where index = $1"#,
            curr_index,
            display_name,
        )
        .execute(&mut txn)
        .await?;
    }

    if let Some(new_index) = new_index {
        sqlx::query!(
            r#"
            update image_tag set index = $2 where index = $1
            "#,
            curr_index,
            new_index
        )
        .execute(&mut txn)
        .await
        .map_err(handle_tag_err)?;
    }

    txn.commit().await?;

    Ok(())
}

pub async fn delete(db: &PgPool, index: i16) -> sqlx::Result<()> {
    sqlx::query!(
        r#"
delete from image_tag where index = $1
            "#,
        index
    )
    .execute(db)
    .await?;

    Ok(())
}
