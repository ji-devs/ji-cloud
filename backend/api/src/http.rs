use std::{net::TcpListener, sync::Arc};

use actix_service::Service;
use actix_web::{
    dev::{MessageBody, Server, ServiceRequest, ServiceResponse},
    error::Error,
    middleware::{Compat, Condition},
    web::{method, Data},
    HttpResponse,
};
use core::{
    config::JSON_BODY_LIMIT,
    env::env_bool,
    http::{get_addr, get_tcp_fd},
    settings::RuntimeSettings,
};
use futures::Future;
use sqlx::postgres::PgPool;
use tracing::Span;
use tracing_actix_web::{root_span, DefaultRootSpanBuilder, RootSpanBuilder, TracingLogger};

use crate::{
    error::BasicError,
    service::{self, mail, s3, upload::cleaner, ServiceData},
    translate,
};

mod cors;
mod endpoints;

fn log_ise<B: MessageBody, T>(
    request: ServiceRequest,
    srv: &T,
) -> impl Future<Output = actix_web::Result<T::Response>>
where
    T: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error>,
{
    let uri: serde_json::Value = request.uri().to_string().into();
    let method: serde_json::Value = request.method().to_string().into();

    let future = srv.call(request);
    async {
        let mut result = future.await?;
        if result.status() == 500 {
            let response: &mut actix_web::HttpResponse<_> = result.response_mut();

            if let Some(err) = response.extensions_mut().remove::<anyhow::Error>() {
                log::error!("ISE while responding to request: {:?}", err);
                sentry::add_breadcrumb(sentry::Breadcrumb {
                    ty: "http".to_owned(),
                    category: Some("request".into()),
                    data: {
                        let mut map = sentry::protocol::Map::new();
                        map.insert("url".to_owned(), uri);
                        map.insert("method".to_owned(), method);
                        map
                    },
                    ..Default::default()
                });

                sentry::integrations::anyhow::capture_anyhow(&err);
            }
        }

        Ok(result)
    }
}

pub struct Application {
    port: u16,
    server: Option<Server>,
}

impl Application {
    pub fn new(port: u16, server: Server) -> Self {
        Self {
            port,
            server: Some(server),
        }
    }

    pub fn port(&self) -> u16 {
        self.port
    }

    // A more expressive name that makes it clear that
    // this function only returns when the application is stopped.
    pub async fn run_until_stopped(mut self) -> Result<(), std::io::Error> {
        if let Some(server) = self.server.take() {
            server.await?;
        }

        Ok(())
    }

    pub async fn stop(mut self, graceful: bool) {
        if let Some(server) = self.server.take() {
            server.stop(graceful).await
        }
    }
}

impl Drop for Application {
    fn drop(&mut self) {
        if let Some(server) = self.server.take() {
            let _ = tokio::spawn(server.stop(false));
        }
    }
}

#[actix_web::main]
pub async fn build_and_run(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<()> {
    let app = build(
        pool,
        settings,
        s3,
        gcp_key_store,
        gcs,
        algolia,
        algolia_key_store,
        jwk_verifier,
        mail_client,
        algolia_manager,
        media_upload_cleaner,
        google_translate,
    )?;
    app.run_until_stopped().await?;

    Ok(())
}

pub fn build(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<Application> {
    let local_insecure = settings.is_local();
    let api_port = settings.api_port;

    let s3 = s3.map(ServiceData::new);
    let gcp_key_store = gcp_key_store.map(ServiceData::new);
    let gcs = gcs.map(ServiceData::new);
    let algolia = algolia.map(ServiceData::new);
    let algolia_key_store = algolia_key_store.map(ServiceData::new);
    let mail_client = mail_client.map(ServiceData::new);
    let algolia_manager = algolia_manager.map(ServiceData::new);
    let media_upload_cleaner = media_upload_cleaner.map(ServiceData::new);
    let google_translate = google_translate.map(ServiceData::new);

    let server = actix_web::HttpServer::new(move || {
        let app = actix_web::App::new()
            .app_data(Data::new(pool.clone()))
            .app_data(Data::new(settings.clone()));

        let app = match s3.clone() {
            Some(s3) => app.app_data(s3),
            None => app,
        };

        let app = match gcp_key_store.clone() {
            Some(gcp_key_store) => app.app_data(gcp_key_store),
            None => app,
        };

        let app = match gcs.clone() {
            Some(gcs) => app.app_data(gcs),
            None => app,
        };

        let app = match algolia.clone() {
            Some(algolia) => app.app_data(algolia),
            None => app,
        };

        let app = match algolia_key_store.clone() {
            Some(algolia_key_store) => app.app_data(algolia_key_store),
            None => app,
        };

        let app = match mail_client.clone() {
            Some(mail_client) => app.app_data(mail_client),
            None => app,
        };

        let app = match algolia_manager.clone() {
            Some(algolia_manager) => app.app_data(algolia_manager),
            None => app,
        };

        let app = match media_upload_cleaner.clone() {
            Some(media_upload_cleaner) => app.app_data(media_upload_cleaner),
            None => app,
        };
        let app = match google_translate.clone() {
            Some(google_translate) => app.app_data(google_translate),
            None => app,
        };

        let enable_tracing_logs = env_bool("ENABLE_TRACING_LOGS");

        app.app_data(Data::from(jwk_verifier.clone()))
            .wrap(cors::get(local_insecure))
            .wrap(Condition::new(
                !enable_tracing_logs,
                Compat::new(actix_web::middleware::Logger::default()),
            ))
            .wrap(Condition::new(
                enable_tracing_logs,
                Compat::new(TracingLogger::<JigziSpanBuilder>::new()),
            ))
            .wrap_fn(log_ise)
            .app_data(
                actix_web::web::JsonConfig::default()
                    .limit(JSON_BODY_LIMIT as usize)
                    .error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::QueryConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::PathConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .external_resource(
                "google_cloud_oauth",
                "https://accounts.google.com/o/oauth2/v2/auth",
            )
            .default_service(actix_web::web::to(default_route))
            .configure(endpoints::user::configure)
            .configure(endpoints::category::configure)
            .configure(endpoints::image::configure)
            .configure(endpoints::audio::configure)
            .configure(endpoints::meta::configure)
            .configure(endpoints::jig::configure)
            .configure(endpoints::module::configure)
            .configure(endpoints::additional_resource::configure)
            .configure(endpoints::jig::report::configure)
            .configure(endpoints::jig::curation::configure)
            .configure(endpoints::course::configure)
            .configure(endpoints::admin::configure)
            .configure(endpoints::animation::configure)
            .configure(endpoints::search::configure)
            .configure(endpoints::media::configure)
            .configure(endpoints::session::configure)
            .configure(endpoints::locale::configure)
            .configure(endpoints::scheduler::configure)
            .configure(endpoints::pdf::configure)
            .configure(endpoints::badge::configure)
            .route("/", method(http::Method::GET).to(no_content_response))
    });

    // if listenfd doesn't take a TcpListener (i.e. we're not running via
    // the command above), we fall back to explicitly binding to a given
    // host:port.
    let listener = if let Some(l) = get_tcp_fd() {
        l
    } else {
        TcpListener::bind(get_addr(Some(api_port)))?
    };

    let port = listener.local_addr().unwrap().port();

    let server = server.listen(listener)?;

    Ok(Application::new(port, server.run()))
}

fn default_route() -> HttpResponse {
    HttpResponse::NotFound().json(BasicError::with_message(
        http::StatusCode::NOT_FOUND,
        "Route not found".to_owned(),
    ))
}

async fn no_content_response() -> HttpResponse {
    HttpResponse::NoContent().finish()
}

pub fn bad_request_handler() -> actix_web::Error {
    BasicError::new(http::StatusCode::BAD_REQUEST).into()
}

pub struct JigziSpanBuilder;

impl RootSpanBuilder for JigziSpanBuilder {
    fn on_request_start(request: &ServiceRequest) -> Span {
        let route = request.match_pattern().unwrap_or("default".to_string());
        let transaction_name = format!("{} {}", request.method(), route);
        let transaction_name = transaction_name.as_str();
        root_span!(request, message = transaction_name)
    }

    fn on_request_end<B>(span: Span, outcome: &Result<ServiceResponse<B>, Error>) {
        DefaultRootSpanBuilder::on_request_end(span, outcome)
    }
}
