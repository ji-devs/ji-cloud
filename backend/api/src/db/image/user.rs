use futures::stream::BoxStream;
use shared::domain::image::{user::UserImage, ImageId, ImageKind};
use sqlx::{PgConnection, PgPool};
use tracing::{instrument, Instrument};
use uuid::Uuid;

#[instrument(skip(pool))]
pub async fn create(pool: &PgPool, user_id: &Uuid, kind: ImageKind) -> sqlx::Result<ImageId> {
    let mut txn = pool.begin().await?;
    let id: ImageId = sqlx::query!(
        //language=SQL
        r#"
insert into user_image_library (user_id, kind)
values ($1, $2)
returning id as "id: ImageId"
"#,
        user_id,
        kind as i16,
    )
    .fetch_one(&mut txn)
    .instrument(tracing::info_span!("inser user_image_library"))
    .await?
    .id;

    sqlx::query!("insert into user_image_upload (image_id) values ($1)", id.0)
        .execute(&mut txn)
        .instrument(tracing::info_span!("inser user_image_upload"))
        .await?;

    txn.commit().await?;

    Ok(id)
}

pub async fn delete(db: &PgPool, user_id: Uuid, image_id: ImageId) -> sqlx::Result<()> {
    sqlx::query!(
        "delete from user_image_library where user_id = $1 and id = $2",
        user_id,
        image_id.0
    )
    .execute(db)
    .await
    .map(drop)
}

pub async fn get(db: &PgPool, user_id: Uuid, image_id: ImageId) -> sqlx::Result<Option<UserImage>> {
    sqlx::query_as!(
        UserImage,
        // language=SQL
        r#"
select id as "id: ImageId", kind as "kind: ImageKind"
from user_image_library
         inner join user_image_upload
                    on user_image_library.id = user_image_upload.image_id
where user_id = $1
  and id = $2
  and processing_result is true
        "#,
        user_id,
        image_id.0,
    )
    .fetch_optional(db)
    .await
}

pub fn list(
    db: &PgPool,
    user_id: Uuid,
    kind: Option<ImageKind>,
) -> BoxStream<'_, sqlx::Result<UserImage>> {
    sqlx::query_as!(
        UserImage,
        // language=SQL
        r#"
select id as "id: ImageId", kind as "kind: ImageKind"
from user_image_library
         join user_image_upload
              on user_image_library.id = user_image_upload.image_id
where processing_result is true
  and user_id = $1
  and (kind is not distinct from $2 or $2 is null)
order by created_at desc
"#,
        user_id,
        kind.map(|it| it as i16)
    )
    .fetch(db)
}

/// checks if the user owns the image requested.
///
/// Returns ResourceNotFound even if the image exists but the user does not h
pub async fn auth_user_image(
    txn: &mut PgConnection,
    user_id: &Uuid,
    image_id: &ImageId,
) -> anyhow::Result<(), crate::error::Upload> {
    let exists = sqlx::query!(
        //language=SQL
        r#"
select exists(select 1 from user_image_library where user_id = $1 and id = $2) as "exists!"
    "#,
        user_id,
        image_id.0
    )
    .fetch_one(txn)
    .await?
    .exists;

    if !exists {
        return Err(crate::error::Upload::ResourceNotFound);
    }

    Ok(())
}
