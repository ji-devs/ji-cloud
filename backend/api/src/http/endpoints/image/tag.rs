//! Manages image tags
use crate::{
    db, error,
    extractor::{ScopeAdmin, TokenUserWithScope},
};
use paperclip::actix::web::Path;
use paperclip::actix::{
    api_v2_operation,
    web::{Data, Json},
    CreatedJson, NoContent,
};
use shared::domain::image::tag::ImageTagResponse;
use shared::domain::meta::ImageTagIndex;
use shared::{
    api::{endpoints, ApiEndpoint},
    domain::image::tag::ImageTagListResponse,
};
use sqlx::PgPool;

#[api_v2_operation]
pub(super) async fn list(
    db: Data<PgPool>,
    _claims: TokenUserWithScope<ScopeAdmin>,
) -> Result<Json<<endpoints::image::tag::List as ApiEndpoint>::Res>, error::NotFound> {
    let image_tags = db::image::tag::list(db.as_ref()).await?;

    Ok(Json(ImageTagListResponse { image_tags }))
}

#[api_v2_operation]
pub(super) async fn create(
    db: Data<PgPool>,
    _claims: TokenUserWithScope<ScopeAdmin>,
    index: Path<ImageTagIndex>,
    req: Json<<endpoints::image::tag::Create as ApiEndpoint>::Req>,
) -> Result<CreatedJson<<endpoints::image::tag::Create as ApiEndpoint>::Res>, error::Tag> {
    let res =
        db::image::tag::create(db.as_ref(), index.into_inner(), req.display_name.as_str()).await?;

    Ok(CreatedJson(ImageTagResponse {
        index: res.0,
        display_name: res.1,
    }))
}

#[api_v2_operation]
pub(super) async fn update(
    db: Data<PgPool>,
    _claims: TokenUserWithScope<ScopeAdmin>,
    index: Path<i16>,
    req: Json<<endpoints::image::tag::Update as ApiEndpoint>::Req>,
) -> Result<NoContent, error::Tag> {
    let req = req.into_inner();

    let _resp = db::image::tag::update(
        db.as_ref(),
        index.into_inner(),
        req.display_name.as_deref(),
        req.index.map(|it| it.0),
    )
    .await?;

    Ok(NoContent)
}

#[api_v2_operation]
pub(super) async fn delete(
    db: Data<PgPool>,
    _claims: TokenUserWithScope<ScopeAdmin>,
    req: Path<i16>,
) -> Result<NoContent, error::Tag> {
    db::image::tag::delete(db.as_ref(), req.into_inner()).await?;

    Ok(NoContent)
}
