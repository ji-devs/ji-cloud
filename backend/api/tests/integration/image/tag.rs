use crate::{
    fixture::Fixture,
    helpers::{initialize_server, LoginExt},
};
use http::StatusCode;
use shared::domain::image::tag::{ImageTagCreateRequest, ImageTagUpdateRequest};

#[actix_rt::test]
async fn create() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image/tag/{}", port, 3))
        .json(&ImageTagCreateRequest {
            display_name: "test name".to_owned(),
        })
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, { ".id" => "[id]" });

    app.stop(false).await;

    Ok(())
}

#[actix_rt::test]
async fn create_conflict() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image/tag/{}", port, 0))
        .json(&ImageTagCreateRequest {
            display_name: "test name".to_owned(),
        })
        .login()
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::CONFLICT);

    app.stop(false).await;

    Ok(())
}

#[actix_rt::test]
async fn list() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    log::info!("making request");

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/image/tag/all", port,))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, { ".**.id" => "[id]" });

    app.stop(false).await;

    Ok(())
}

async fn update(index: i16, req: ImageTagUpdateRequest) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!("http://0.0.0.0:{}/v1/image/tag/{}", port, index,))
        .json(&req)
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/image/tag/all", port,))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, { ".**.id" => "[id]" });

    app.stop(false).await;

    Ok(())
}

#[actix_rt::test]
async fn update_no_index() -> anyhow::Result<()> {
    update(
        0,
        ImageTagUpdateRequest {
            display_name: Some("test".to_owned()),
            index: None,
        },
    )
    .await
}

#[actix_rt::test]
async fn update_with_index() -> anyhow::Result<()> {
    update(
        1,
        ImageTagUpdateRequest {
            display_name: Some("test".to_owned()),
            index: Some(15),
        },
    )
    .await
}

#[actix_rt::test]
async fn update_none() -> anyhow::Result<()> {
    update(
        1,
        ImageTagUpdateRequest {
            display_name: None,
            index: None,
        },
    )
    .await
}

#[actix_rt::test]
async fn update_only_index() -> anyhow::Result<()> {
    update(
        1,
        ImageTagUpdateRequest {
            display_name: None,
            index: Some(3),
        },
    )
    .await
}

#[actix_rt::test]
async fn update_conflict() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!("http://0.0.0.0:{}/v1/image/tag/{}", port, 1))
        .json(&ImageTagUpdateRequest {
            display_name: None,
            index: Some(0),
        })
        .login()
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::CONFLICT);

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/image/tag/all", port,))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, { ".**.id" => "[id]" });

    app.stop(false).await;

    Ok(())
}

#[actix_rt::test]
async fn delete() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Image, Fixture::MetaKinds], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .delete(&format!("http://0.0.0.0:{}/v1/image/tag/2", port,))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    app.stop(false).await;

    Ok(())
}
