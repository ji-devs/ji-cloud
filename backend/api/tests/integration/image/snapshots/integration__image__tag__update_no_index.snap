---
source: tests/integration/image/tag.rs
expression: body

---
{
  "image_tags": [
    {
      "index": 0,
      "display_name": "test"
    },
    {
      "index": 1,
      "display_name": "B"
    },
    {
      "index": 2,
      "display_name": "C"
    }
  ]
}
